File: app/src/main/java/dev/favourdevlabs/cleanthes/ui/base/AuthenticatedActivity.java
package dev.favourdevlabs.cleanthes.ui.base;

import android.content.Intent;
import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import dev.favourdevlabs.cleanthes.ui.auth.LoginActivity;
import dev.favourdevlabs.cleanthes.ui.auth.SessionManager;

public abstract class AuthenticatedActivity extends AppCompatActivity {

	@Override
	protected void onStart() {
		super.onStart();
		if (!SessionManager.isUnlocked()) {
			redirectToLogin();
		}
	}

	@Override
	protected void onResume() {
		super.onResume();
		// Only refresh — never make redirect decisions here
		if (SessionManager.isUnlocked()) {
			SessionManager.refreshSession();
		}
	}

	protected void redirectToLogin() {
		Intent intent = new Intent(this, LoginActivity.class);
		intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
		startActivity(intent);
		finish();
	}
}

File: app/src/main/java/dev/favourdevlabs/cleanthes/ui/base/AuthenticatedActivity.kt
package dev.favourdevlabs.cleanthes.ui.base

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import dagger.hilt.android.AndroidEntryPoint
import dev.favourdevlabs.cleanthes.security.session.SessionManager
import dev.favourdevlabs.cleanthes.ui.auth.LoginActivity
import javax.inject.Inject

@AndroidEntryPoint
abstract class AuthenticatedActivity : AppCompatActivity() {
    @Inject lateinit var sessionManager: SessionManager

    override fun onStart() {
        super.onStart()
        if (sessionManager.lockState.value) {
            redirectToLogin()
        }
    }

    override fun onResume() {
        super.onResume()
        if (!sessionManager.lockState.value) {
            sessionManager.refreshSession()
        }
    }

    protected fun redirectToLogin() {
        val intent =
            Intent(this, LoginActivity::class.java).apply {
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
            }
        startActivity(intent)
        finish()
    }
}
